play/src/lib/HelperFunctions.tsx

```diff
@@ -129,4 +130,33 @@ export const fetchLocalData = () => {
   if (!localModel)
     window.localStorage.removeItem(StorageKey);
   return localModel;
+};
+
+/**
+ * sorts teams by score descending, then alphabetically by name
+ * only include teams with scores in the top five
+ * @param inputTeams - the teams to be sorted
+ * @returns - the sorted teams
+ */ 
+export const teamSorter = (inputTeams: ITeam[]) => {
+    // create a set to store unique scores with no repeats
+    const scoreSet = new Set<number>();
+    inputTeams.forEach((team) => scoreSet.add(team.score));
+    // convert the set to an array and sort it in descending order to retrieve only the top five highest scores
+    const sortedTopScores: Set<number> = new Set(Array.from(scoreSet)
+      .sort((a, b) => b - a)
+      .slice(0, 5));
+    // Filter through sortedTeams for all teams with the top five unique scores
+    const topTeamsUnsorted: ITeam[] = inputTeams.filter((team) =>
+      sortedTopScores.has(team.score)
+    );
+    const topTeamsSorted: ITeam[] = topTeamsUnsorted.sort((a, b) => {
+      if (a.score !== b.score) {
+        // sort by score descending
+        return b.score - a.score;
+      }
+      // sort alphabetically by name if scores are tied
+      return a.name.localeCompare(b.name);
+    });
+    return topTeamsSorted;
```

Comment:
Although the above code works, but it's not efficient, it's iterating the teams multiple times. Being efficient for this part isn't that much important in our use-case as we have small number of teams, but if the number of teams increases significantly, modifying this to more efficient algorithm would be necessary. I'm going to use this as an opportunity to discuss some boring stuff about algorithms and runtime so feel free to ignore if not interested in them.

One thing that you might be interested in reading on is Insertion sort algorithm. Basically what it does is keeping the array sorted when a new items are added. This algorithm runtime is `O(n^2)` which makes it not efficient when the data is large, but for small numbers, IIRC arrays of size up to 20, this is an efficient algorithm to keep your array sorted while adding new items to it. This could be one approach to use here as our team size is less than 20. However, if we scale to support more teams, this wouldn't be the most efficient way of doing it.

You can take advantage of existing `sort`, `reverse` and the fact that in javascript `0` means `false` and anything else means `true`:

```typescript
inputTeams.sort((lhs, rhs) => {
    const scoreDiff = lhs.score - rhs.score
    if (scoreDiff === 0) {
        return lhs.name.localeCompare(rhs.name);
    }
    return scoreDiff
}).reverse().slice(0, 5)
```

one-liner version:

```typescript
const result = arr.sort((lhs, rhs) => {
     return lhs.score - rhs.score ?? lhs.name.localeCompare(rhs.name);
}).reverse().slice(0, 3)
```

Reply:
cc @josephinaim @Alexandra-Kushnirsky in case if you're interested in the above details

Reply:
Super cool explanation here and thanks for the code samples - those are so clean. For some context, this function is used to sort the Leaderboard at the end of a game. The requirements of that leaderboard are as follows:

1. Only show teams with a score in the top 5
2. If there are no ties, only 5 teams will be displayed.
3. If there are tied teams within this top 5, sort those tied teams alphabetically. More than 5 teams can be displayed. 

Your code is similar to what Josephina initially had for the sorting [here](https://github.com/rightoneducation/righton-app/pull/690/files) but unfortunately, we also need to ensure that we're including all possible teams within the top 5. 

Josephina and I were discussing this yesterday and this morning. Here's some of our thought process: https://github.com/rightoneducation/righton-app/pull/710. See my comment regarding the ordering of the sorts to cap the length of the search.

Basically, if we're starting with an unordered array, we need to sort the teams by score. Then, within that group of top five scores, we need to find the ties and sort each of those subsets alphabetically. Both Josephina and I were struggling with how to consolidate those two requirements into a single pass. What we can up with was to generate a set of unique scores, sort those, and then filter the existing array by that set. Then we could sort any ties within that limited set alphabetically. 

It would be nice if we could just count the number of ties and adjust the slice accordingly, like so:

```export const teamSorter = (inputTeams: ITeam[]) => {
  let ties = 0;
  return inputTeams.sort((lhs, rhs) => {
    const scoreDiff = lhs.score - rhs.score
    if (scoreDiff === 0) {
        ties =+ 1;
        return lhs.name.localeCompare(rhs.name);
    }
    return scoreDiff
  }).reverse().slice(0, 5 + ties);
};
```

But this blows up, because it is counting every tie (so if the top five has no ties and the bottom five are all tied, this will show the entire leaderboard). 

We were struggling with how to get around this issue and try to make it in a single pass. Do you have any other ideas to incorporate that final requirement?

Also, for my own education, could you explain how our original function is O(N^2)? I was thinking it was O(N) because we :

1. forEach - O(N)
2. sort - O(N)
3. filter.has - O(N) + O(1)
4. sort - O(N)

simplifies to O(N) (worst case: every team is in the top five, in some arrangement of ties).

I'm definitely still wrapping my head around Big O notation in practice so would you be able to steer me in where I'm going wrong on this? Thanks a ton for the comments @maniramezan!

Reply:
aww, sorry I missed that part, for it you can modify the above to something like:

```typescript
const newSortTeams = (teams: ITeam[]) => {
    const sortedTeams = teams.sort((lhs, rhs) => {
        return lhs.score - rhs.score ?? lhs.name.localeCompare(rhs.name);
   }).reverse()
   
   var lastScore = -1
   let totalTeams = 5
   let ret = Array(totalTeams)
   sortedTeams.every((team) => {
       if (team.score !== lastScore) {
           totalTeams -= 1
       }
       ret.push(team)
       return totalTeams !== 0
   })
   
   return ret
}
```

So the sort BigO is usually considered to be `nlogn` as mostly  they use quicksort. So generally, the proposed solution and the current solution both have the same BigO notation of `nlogn`, but memory footprint and also the actual performance would be different when you use a large set of data. 

So I started being curious about the actual performance comparison, so I created [this fiddle](https://jsfiddle.net/21xrd80y/2/). It looks like actually my proposed solution performed worse,  🤦‍♂️. I was pretty sure it'd be due to `reversed` method not being performant and doing some copy / iterating over the array work hence I modified the proposed solution to the following:

```typescript
const newNewSortTeams = (teams: ITeam[]) => {
    const sortedTeams = teams.sort((lhs, rhs) => {
        return lhs.score - rhs.score ?? lhs.name.localeCompare(rhs.name);
   })
   
   var lastScore = -1
   let totalTeams = 5
   let ret = Array(totalTeams)
   for (var i = sortedTeams.length - 1; i >=0 && totalTeams > 0; i--) {
       if (sortedTeams[i].score !== lastScore) {
           totalTeams -= 1
       }
       ret.push(sortedTeams[i])
   }
   
   return ret
}
```

And now, the numbers for running all three implementations for 10000000 are as follow:

```
sortTeams: 685.921875 ms
newSortTeams: 1706.89306640625 ms
newNewSortTeams: 618.531982421875 ms
```

As also mentioned before, the current implementation would suffice for current and any foreseeable version of the app. I was personally just curious on how this could be improved.

Reply:
This is so interesting. Wild that the plain for loop seems like the go to here. I'm signing off for the weekend but want to take a look at the fiddle and see what's going on. I'll probably just implement your revised though, as it seems both faster and easier to read. Also, thanks for the O(nlogn) explanation. totally makes sense with quicksort.

Reply:
I decided to go with your for loop implementation here to save a few ms and because it seems cleaner (doesn't need sets etc).